### ISSUE TITLE
Consider updating NaCl as an alternative to WebAssembly

### ISSUE THREAD

**Issue by slipher** [DaemonEngine/Daemon#795]
Google Native Client, or NaCl, is not as dead as has been previously believed. Although support for NaCl in websites and 3rd-party browser extensions has been removed, it lives on in some form. [This bug tracking issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1106311) details the development in 2020-2021 of a C++17-capable NaCl compiler toolchain, codenamed "Saigo". CI-related commits show that it is actively tested on numerous platforms, so it is not just a ChromeOS thing.

After poking around a bit in [Chromium Code Search](https://source.chromium.org/), I still haven't been able to determine what NaCl is being used for. Chromium has a lot of "sandboxed" processes, but in general these are not NaCl-based. But as of now the code is being shipped with Google Chrome: inspection of a Windows install shows a `nacl_irt_x86_64.nexe` and various NaCl-related strings in `chrome.dll`. So the chances of it being available for any platform Daemon would want to support seem good. (Contrast this with standalone WebAssembly engines which are more popular for Linux servers and have not been ported to 32-bit architectures.)

**Comment by illwieckz** [comment]
Some weeks ago I investigated “chromium builds” and only Linux had nacl stuff. Even macOS amd64 chromium did not have any nacl stuff, and I have seen some commits disabling nacl for macOS.

**Comment by illwieckz** [comment]
Some search about “Saigo” and “Native Client” returned me this:

https://chromium.googlesource.com/native_client/nacl-llvm-project-v10/+/refs/heads/saigo-2023-01-02

**Comment by necessarily-equal** [comment]
For the record, the C++ toolchain that we would be most likely to use for wasm is upstream clang. This might be the lowest maintenance toolchain we could ever get.

**Comment by DolceTriade** [comment]
+1, if Google somehow makes the NaCl thing slightly more use friendly, I think it's better than wasm for now. Wasm doesn't really provide us any features beyond even more modern C++.

For example, in the switch from QVM -> NaCl, if we could have supported C++ on QVMs, we probably would have done that instead of using NaCl.

**Comment by illwieckz** [comment]
To be able to upgrade NaCl we need two things:

- a working SDK able to build nexe for hardware platforms we want to run the game on,
  * at least, the SDK should run on Linux amd64 to be able to build and release the game, but to be on par with current SDK such SDK should run on Linux, Windows and macOS amd64;
- a working runner to ship with the game for hardware platforms and operating systems we want to run the game on,
  * basically the `nacl_loader` binary and the `irt` nexe thing.

**Comment by bmorel** [comment]
I think it is also important to reference [this](https://forums.unvanquished.net/viewtopic.php?p=19109&sid=50da15436e177c20c525851d4bb0daff#p19109) thread on the forum, which says that wasm is 2 times slower than native binaries and author seems to have done their research there.

Performance is important as people can only rent cheap VPS to host servers with their own gameplay, and test their own mods. Bots are notably performance hungry, and they were moved from engine to gamelogic **thanks** to pNaCl near-native performances, I suppose.
On my VPS, instances running with 10 bots were (were, because I need to update my codebase) in average using 10% CPU. If it is true that WASM is roughly twice as slow, this might imply going up to 20% CPU, for just 10 bots.
I personally do not intend to spend more money in a beefier server just to run servers, so performance are a concern to me. I did not started renting this machine for unvanquished, and unvanquished already have the lion's share of performances, which is ok as it does not impact other processes, but such an increase might very well change that fact.

Keeping up with pNaCl, we would avoid spending developer time (a very scarce resource) on another porting, everything is already there. We would avoid hardly avoidable regressions, too, that can be very annoying. Avoiding compatibility break and sync problems between clients and servers, too.

The main incentive to move to WASM was, IIRC, that pNaCl was thought dead. If it is not, then it would probably be a waste of time, except if the move is almost finished, in which case the decision is a bit harder to do, I guess.

**Comment by illwieckz** [comment]
About updating nacl, there are some old-but-less-old versions available:

```sh
$ curl -s https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/naclsdk_manifest2.json | grep -E '"url":|"description":|"sha1":|"stability":|"name":' | sed -e 's/"//g;s/,$//;s/^      //' | tac
stability: canary
name: pepper_canary
description: Chrome 58 bundle. Chrome revision: 447822. NaCl revision: 70540f6583440d4174e6d75743bb544cfbaaf204
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/trunk.447822/naclsdk_win.tar.bz2
      sha1: 183a705dd97ccb9927d53539deb27fb58a9bb611
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/trunk.447822/naclsdk_mac.tar.bz2
      sha1: 645de9e81a14801245d5954d71274407d6537f73
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/trunk.447822/naclsdk_linux.tar.bz2
      sha1: cb072ab68302803dd4ccf61346db274f3cb5422d
stability: dev
name: pepper_56
description: Chrome 56 bundle. Chrome revision: 54. NaCl revision: 163dfeb43e76995b4265ecd4e78670f7dd432e44
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/56.0.2924.3/naclsdk_win.tar.bz2
      sha1: f20db79c123ef90a046d5399c1e86eff2ca117ed
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/56.0.2924.3/naclsdk_mac.tar.bz2
      sha1: af32526bbde13e60905bbe03a834a8a97daaed92
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/56.0.2924.3/naclsdk_linux.tar.bz2
      sha1: 4f9d6402bc645e137fd601ede6eec67e809e7d6a
stability: beta
name: pepper_55
description: Chrome 55 bundle. Chrome revision: 624. NaCl revision: fe5bd51f9b79d27b2a3884181d4004cfccf70759
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/55.0.2883.59/naclsdk_win.tar.bz2
      sha1: d1f4dc70ceea1dea40e0d1cbbb96d35476ccf681
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/55.0.2883.59/naclsdk_mac.tar.bz2
      sha1: a4bb07075a8509b07e4ff74564ce4632b79dda48
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/55.0.2883.59/naclsdk_linux.tar.bz2
      sha1: 3453c02973381ef5bea50e1eaa3ac2c5418f79d8
stability: beta
name: pepper_50
description: Chrome 50 bundle. Chrome revision: 250. NaCl revision: d182b0a028d44ba9767ae25b30cf3986a7e72e99
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/50.0.2661.37/naclsdk_win.tar.bz2
      sha1: baf5d2c2fccb50bc015dd57c66ccff46514d66a1
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/50.0.2661.37/naclsdk_mac.tar.bz2
      sha1: f306e86a0da969f87ff0637d6786258da31d006e
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/50.0.2661.37/naclsdk_linux.tar.bz2
      sha1: 356ed3c41abe6d062a9f357138c34a352a67a17a
stability: stable
name: pepper_49
description: Chrome 49 bundle. Chrome revision: 592. NaCl revision: 26c7e6ae1327a43c948f4666fef9e84fe351ab7c
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/49.0.2623.87/naclsdk_win.tar.bz2
      sha1: 368b4e141495640e483563d53cf8885b2f6d2a7e
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/49.0.2623.87/naclsdk_mac.tar.bz2
      sha1: 87dcc561b8a77a17c6bbdce9de501ce1c3afc7c2
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/49.0.2623.87/naclsdk_linux.tar.bz2
      sha1: c0085301e5fc2357f7040054d24a140be70f4d2d
stability: post_stable
name: pepper_47
description: Chrome 47 bundle. Chrome revision: 545. NaCl revision: 719fce7fa0c1f11cf05d5c20cdfa6648e8a35611
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/47.0.2526.111/naclsdk_win.tar.bz2
      sha1: 5ed9c96786eae063f6931e586e7a345193fed56e
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/47.0.2526.111/naclsdk_mac.tar.bz2
      sha1: 402e2b71aae904ed7e780602381aa6316f3664a6
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/47.0.2526.111/naclsdk_linux.tar.bz2
      sha1: 43b593f28342ddbbe0a6d1b4b22ba6b5d62a2fd3
stability: dev
name: vs_addin
description: Visual Studio Add-In for building and debugging NaCl applications
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/sdk/1578/vs_addin.tgz
      sha1: c344205faee543e058a978ed1c06cdf5cff248c6
stability: stable
name: sdk_tools
description: Native Client SDK Tools, revision 361478
    url: https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/trunk.361478/sdk_tools.tgz
      sha1: c90944aa6dc45085f1a222396ffa2b6a86afdf3d
```

We are currently using `pepper44` (2015-08-08) which is not listed in the manifest:

```
https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_win.tar.bz2
https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_mac.tar.bz2
https://storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_linux.tar.bz2
```

```
$ sha1sum -b storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_*.tar.bz2
40c3aea429d262900b5b1453ed8955865a90599a *storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_linux.tar.bz2
f46e4b263f25af9ae9011d2ec5860f8b0baef825 *storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_mac.tar.bz2
11c65e7703aba1d20e18f6ed28fecd704a2fb507 *storage.googleapis.com/nativeclient-mirror/nacl/nacl_sdk/44.0.2403.155/naclsdk_win.tar.bz2
```

The latest public `stable` release  was `pepper_49` (2016-03-08), the most recent public `canary` release is `pepper_canary` which would have been expected to be `pepper_58` (2017-02-02).

**Comment by illwieckz** [comment]
For reference, we use a `pnacl-clang` based on `3.6.0` while we can use one based on `3.7.0`:

```
$ PNACLPYTHON=python2 pepper_44/toolchain/linux_pnacl/bin/pnacl-clang++ --version
clang version 3.6.0 (https://chromium.googlesource.com/a/native_client/pnacl-clang.git 96b3da27dcefc9d152e51cf54280989b2206d789) (https://chromium.googlesource.com/a/native_client/pnacl-llvm.git d0089f0b008e03cfd141f05c80e3b628c2df75c1) nacl-version=0beb554650e57834a9fdda91ca9e3352b330b653
Target: le32-unknown-nacl
Thread model: posix

$ PNACLPYTHON=python2 pepper_49/toolchain/linux_pnacl/bin/pnacl-clang++ --version
clang version 3.7.0 (https://chromium.googlesource.com/a/native_client/pnacl-clang.git cf0dc7f6e6123dfa9b8834b56743315300b34e6c) (https://chromium.googlesource.com/a/native_client/pnacl-llvm.git baa63524b6b493ec2a6aa2c5193d9f25c0c33191) nacl-version=3f81c4b7f4cc2bde2ac49873d9d666c02206d459
Target: le32-unknown-nacl
Thread model: posix

$ PNACLPYTHON=python2 pepper_58/toolchain/linux_pnacl/bin/pnacl-clang++ --version
clang version 3.7.0 (https://chromium.googlesource.com/a/native_client/pnacl-clang.git ce163fdd0f16b4481e5cf77a16d45e9b4dc8300e) (https://chromium.googlesource.com/a/native_client/pnacl-llvm.git 7251d5b59fca15195c94a3a7da70f0081724448f) nacl-version=5dfe030a71ca66e72c5719ef5034c2ed24706c43
Target: le32-unknown-nacl
Thread model: posix
```

**Comment by slipher** [comment]
I tried to upgrade to the latest release (which was like 6 months newer than the one we're using) one time, but abandoned the attempt because it turned out that toolchain support for C++ exceptions was removed. It's not that hard for us to live without exceptions, but it just didn't seem worthwhile since the benefits were too small.

**Comment by illwieckz** [comment]
Well…

```
toolchain/linux_x86/nacl_x86_newlib/bin/x86_64-nacl-g++ -v
Utilisation des specs internes.
Cible : x86_64-nacl
Configuré avec: /mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/SRC/gcc/configure CC='gcc -m32' CFLAGS='-O2 -g -Dinhibit_libc -D__gthr_posix_h' LDFLAGS=-s 'CFLAGS_FOR_TARGET=-O2 -g -I/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/BUILD/headers_for_build' 'CXXFLAGS_FOR_TARGET=-O2 -g -I/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/BUILD/headers_for_build' --prefix=/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/sdk/nacl-sdk --disable-decimal-float --disable-libgomp --disable-libmudflap --disable-libssp --disable-libstdcxx-pch --target=x86_64-nacl --enable-linker-build-id --with-gmp=/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/BUILD/.gcc-extra-install-gmp --with-mpfr=/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/BUILD/.gcc-extra-install-mpfr --with-ppl=/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/BUILD/.gcc-extra-install-ppl --with-host-libstdcxx='-lpwl -Wl,-Bstatic -lstdc++ -Wl,-Bdynamic -lm' --with-cloog=/mnt/data/b/build/slave/nacl-toolchain/build/native_client/tools/BUILD/.gcc-extra-install-cloog-ppl --disable-ppl-version-check --disable-shared --enable-languages=c,c++,objc --enable-threads=nacl --enable-tls --with-newlib
Modèle de thread: nacl
gcc version 4.4.3 20150409 (Native Client 11953a35fcd4f999b097a343bc49d89f94af21a8, Git Commit f80d6b9ee7f94755c697ffb7194fb01dd0c537dd) (GCC) 
```

The build is from 2015… :-/

**Comment by illwieckz** [comment]
The `pnacl_clang+`, it's clang 3.7.0 and the build seems to be dated from 2022-04-01:

```
toolchain/linux_x86/pnacl_newlib/bin/pnacl-clang++ -v
clang version 3.7.0 (https://chromium.googlesource.com/a/native_client/pnacl-clang.git c029fe2e0250ec34ba325624ab659835f06bf777) (https://chromium.googlesource.com/a/native_client/pnacl-llvm.git a3b836a9847468b1be6247bd596ee12d5783e332)
Target: le32-unknown-nacl
Thread model: posix
clang: warning: argument unused during compilation: '-O0'
clang: warning: argument unused during compilation: '-fno-vectorize'
clang: warning: argument unused during compilation: '-fno-slp-vectorize'
clang: warning: argument unused during compilation: '-fno-common'
clang: warning: argument unused during compilation: '-nostdinc'
clang: warning: argument unused during compilation: '-fno-gnu-inline-asm'
clang: warning: argument unused during compilation: '-fexceptions'

stat -c '%y' toolchain/linux_x86/pnacl_newlib/bin/pnacl-clang++
2022-04-01 19:17:46.186732530 +0200
```

**Comment by illwieckz** [comment]
I managed to get saigo build an `helloworld.nexe` that runs, but `exception.nexe` doesn't.

This is based on clang 17 and the build looks to be from today:

```
$ toolchain/linux_x86/saigo_newlib/bin/x86_64-nacl-clang++ helloworld.cpp -o helloworld.nexe
$ ./run.py helloworld.nexe 
********************************************************************************
                        helloworld.nexe is X86-64 STATIC                        
********************************************************************************

/home/illwieckz/dev/native_client/depot/native_client/scons-out/opt-linux-x86-64/staging/nacl_helper_bootstrap /home/illwieckz/dev/native_client/depot/native_client/scons-out/opt-linux-x86-64/staging/sel_ldr --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B /home/illwieckz/dev/native_client/depot/native_client/scons-out/nacl_irt-x86-64/staging/irt_core.nexe -- /home/illwieckz/dev/native_client/depot/native_client/helloworld.nexe

DEBUG MODE ENABLED (bypass acl)
[1850944,3838008192:13:51:49.826304] BYPASSING ALL ACL CHECKS
[1850944,3838008192:13:51:49.826358] Native Client module will be loaded at base address 0x0000631600000000
[1850944,3838008192:13:51:49.826365] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
[1850944,3838008192:13:51:49.827730] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
[1850944,3838008192:13:51:49.827736] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
hello world

$ toolchain/linux_x86/saigo_newlib/bin/x86_64-nacl-clang++ exception.cpp -o exception.nexe
$ ./run.py exception.nexe 
********************************************************************************
                        exception.nexe is X86-64 STATIC                        
********************************************************************************

/home/illwieckz/dev/native_client/depot/native_client/scons-out/opt-linux-x86-64/staging/nacl_helper_bootstrap /home/illwieckz/dev/native_client/depot/native_client/scons-out/opt-linux-x86-64/staging/sel_ldr --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B /home/illwieckz/dev/native_client/depot/native_client/scons-out/nacl_irt-x86-64/staging/irt_core.nexe -- /home/illwieckz/dev/native_client/depot/native_client/exception.nexe

DEBUG MODE ENABLED (bypass acl)
[1851531,2455493504:13:52:11.454804] BYPASSING ALL ACL CHECKS
[1851531,2455493504:13:52:11.454862] Native Client module will be loaded at base address 0x0000303200000000
[1851531,2455493504:13:52:11.454868] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
[1851531,2455493504:13:52:11.464903] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
[1851531,2455493504:13:52:11.464911] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK

** Signal 11 from untrusted code: pc=3032fffeff20

$ toolchain/linux_x86/saigo_newlib/bin/x86_64-nacl-clang++ -v
clang version 17.0.0 (https://chromium.googlesource.com/a/native_client/nacl-llvm-project-v10.git f1bbea6e081a4d53853028aa51c5963a8b04feb4)
Target: x86_64-unknown-nacl
Thread model: posix
InstalledDir: /home/illwieckz/dev/native_client/depot/native_client/toolchain/linux_x86/saigo_newlib/bin

$ stat -c '%y' toolchain/linux_x86/saigo_newlib/bin/x86_64-nacl-clang++
2023-03-29 13:07:14.211783585 +0200
```

**Comment by illwieckz** [comment]
That saigo clang doesn't complain if I use `-std=c++20`:

```
$ toolchain/linux_x86/saigo_newlib/bin/x86_64-nacl-clang++ -std=c++20 helloworld.cpp -o helloworld.nexe

$ ./run.py helloworld.nexe 
********************************************************************************
                        helloworld.nexe is X86-64 STATIC                        
********************************************************************************

/home/illwieckz/dev/native_client/depot/native_client/scons-out/opt-linux-x86-64/staging/nacl_helper_bootstrap /home/illwieckz/dev/native_client/depot/native_client/scons-out/opt-linux-x86-64/staging/sel_ldr --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B /home/illwieckz/dev/native_client/depot/native_client/scons-out/nacl_irt-x86-64/staging/irt_core.nexe -- /home/illwieckz/dev/native_client/depot/native_client/helloworld.nexe

DEBUG MODE ENABLED (bypass acl)
[1859664,266750848:13:58:36.653656] BYPASSING ALL ACL CHECKS
[1859664,266750848:13:58:36.653708] Native Client module will be loaded at base address 0x00007c8300000000
[1859664,266750848:13:58:36.653716] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
[1859664,266750848:13:58:36.655072] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
[1859664,266750848:13:58:36.655078] WARNING: BYPASSING DESCRIPTOR SAFETY CHECK
hello world
```

**Comment by illwieckz** [comment]
while `run.py` automatically builds the loader if it's missing, this command can do it as well:

```
./scons platform=x86-64 MODE=opt-host naclsdk_validate=0 sysinfo=0 -j8 sel_ldr
```

**Comment by illwieckz** [comment]
Even with a saigo toolchain from today there is still no `arm64` target:

```
$ toolchain/linux_x86/saigo_newlib/bin/clang++ --print-targets

  Registered Targets:
    arm     - ARM
    armeb   - ARM (big endian)
    thumb   - Thumb
    thumbeb - Thumb (big endian)
    x86     - 32-bit X86: Pentium-Pro and above
    x86-64  - 64-bit X86: EM64T and AMD64


toolchain/linux_x86/saigo_newlib/bin/clang++ --target=arm64--nacl helloworld.cpp -o helloworld.nexe
clang++: error: the target architecture 'arm64' is not supported by the target 'Native Client'
```

**Comment by illwieckz** [comment]
Since it's doable to rebuild the nacl_loader, out of curiosity I wonder if it can be rebuilt for armel (and then, Android).

**Comment by slipher** [comment]
If the PNaCl format no longer exists, that may or may not be bad for us... yes, we would have to compile all the code multiple times, but the translation (PNaCl -> NaCl) step is so slow that the overall build time might not really be worse. I bet it's faster at least for incremental builds.

**Comment by DolceTriade** [comment]
Sounds like a worthwhile thing to test then

**Comment by illwieckz** [comment]
I will probably spend time on updating NaCl only if we find [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime) unusable, see:

 - https://github.com/DaemonEngine/Daemon/issues/757#issuecomment-1769041733
 
I consider that the priority on that topic after next release is to investigate WAMR then fallback on updated NaCl if WAMR is proven unusable.

**Comment by bmorel** [comment]
If work is done on "moving", please keep in consideration the need to have a transition period where *both* are supported, as such a move can trigger hard to find bugs in mods, that only time can reveal.

**Comment by illwieckz** [comment]
I assume the way to build for a newer NaCl or for Wasm would require a very different CMake code than our current code with PNaCL, so I think that would even be easier to not modify the existing code but to add code, this means having the two toolchains in CMake may be even easier. For the virtual machine plumbing I have no idea though.

At QVM/NaCl time we did a transition where both existed at the same time, and there was a time were the server was running NaCl by default and the client was still running QVM by default while ironing out the last bits.

**Comment by slipher** [comment]
Does Saigo support setjmp/longjmp at least?

**Comment by illwieckz** [comment]
I have no idea about setjmp and Saigo!

The instructions above are enough to get a working compiler on Linux, which makes it possible to investigate saigo with PoC or even benchmarks, and test for specific feature support. For a Linux developer everything is available, it's just not really designed to be redistributed.

For more information about a possible migration to saigo, it is very easy to build the saigo compiler itself (it's just very hungry on resources): it's just like building clang since it's a clang fork.

What is complicated is to build the nacl libc/libc++ and maybe some other tools. A first migration step would be to make it possible to build the saigo compiler but rely on prebuilt saigo libc/libc++.

**Comment by slipher** [comment]
setjmp/longjmp seem to work with some [simple examples](http://groups.di.unipi.it/~nids/docs/longjump_try_trow_catch.html).

**Comment by slipher** [comment]
illwieckz is implementing this, thanks! New tracking issue: https://github.com/Unvanquished/Unvanquished/issues/3197

**Comment by illwieckz** [comment]
Funnily, I discovered there is a “Samsung NaCl SDK” meant to be used with “Samsung Studio” for development for Samsung Smart TVs…

It is said that “Samsung has customized the NaCl technology and has supported it on Smart TVs since 2013” and that “Tizen TV will continue its support for NaCl only until 2021-year products”:
https://developer.samsung.com/smarttv/develop/extension-libraries/nacl/overview.html

Their Pepper version (63) is higher than the latest Pepper canary distributed by Google (58):
https://developer.samsung.com/smarttv/develop/extension-libraries/nacl/download.html

The files in the archive are from 2019-02-12, but the provided PNaCl clang is 3.7.0 and the NaCl GCC is 4.4.3.

So, we miss nothing anyway. Saigo is way much better.
